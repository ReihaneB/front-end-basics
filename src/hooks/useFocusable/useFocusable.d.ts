export interface UseFocusableProps {
  autoFocus?: boolean;
}
