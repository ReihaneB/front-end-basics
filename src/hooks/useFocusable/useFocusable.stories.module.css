.focus {
  &:focus {
    background-color: gold;

    &::after {
      content: ' focused';
    }
  }
}

.focusable {
  all: initial;
  display: block;
  background-color: silver;

  &::after {
    content: ' not focused';
  }
  
  composes: focus;
}
