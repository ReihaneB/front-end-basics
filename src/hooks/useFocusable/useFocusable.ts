import { useRef, useEffect, useImperativeHandle, Ref, useCallback, RefObject, LegacyRef } from 'react';

import { UseFocusableProps } from './useFocusable.d';

export function useFocusable(
  ref: Ref<unknown>,
  { autoFocus = false }: UseFocusableProps
) {
  const elementRef = useRef(null) as RefObject<HTMLButtonElement> & LegacyRef<HTMLDivElement> & Ref<HTMLAnchorElement>;

  useEffect(() => {
    if (autoFocus && elementRef.current) {
      elementRef.current.focus({ preventScroll: true });
    }
  }, [autoFocus]);

  useImperativeHandle(ref, () => ({
    focus: () => {
      if (elementRef.current) {
        elementRef.current.focus();
      }
    },
  }));

  const setFocus = useCallback(() => {
    if (elementRef.current) {
      elementRef.current.focus({ preventScroll: true });
    }
  }, []);

  return {
    ref: elementRef,
    setFocus,
  };
}
