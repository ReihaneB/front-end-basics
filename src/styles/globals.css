:root {
  /* colors */
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  border: 0;
  font: inherit;
  vertical-align: baseline;
}

html,
body {
  max-width: 100vw;
  overflow-x: hidden;
}

details,
figcaption,
figure,
footer,
hgroup,
menu,
nav {
  display: 'block',
}

ol,
ul {
  list-style: none;
}

blockquote,
q {
  quotes: none;
}

blockquote::before,
blockquote::after,
q::before,
q::after {
  content: '';
  content: none;
}

table {
  border-collapse: collapse;
  border-spacing: 0;
}

a {
  color: inherit;
  text-decoration: none;
}

h1 {
  font-size: 30px;
  line-height: 38px;
  font-weight: bold;
}

h2 {
  font-size: 20px;
  line-height: 24px;
  font-weight: bold;
}

h3 {
  font-size: 16px;
  line-height: 20px;
  font-weight: bold;
}

h4 {
  font-size: 14px;
  line-height: 18px;
}

h5 {
  font-size: 12px;
  line-height: 16px;
}

p {
  font-size: 16px;
  line-height: 24px;
}