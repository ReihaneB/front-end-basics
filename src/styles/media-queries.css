@custom-media --small-mobile (max-width: 375px);
@custom-media --mobile (max-width: 767px);
@custom-media --small-tablet (max-width: 957px);
@custom-media --tablet (max-width: 1199px);
@custom-media --ultra-wide (min-width: 1920px);