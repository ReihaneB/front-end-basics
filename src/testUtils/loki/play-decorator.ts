/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-expect-error
import createAsyncCallback from '@loki/create-async-callback';
// @ts-expect-error
import isLokiRunning from '@loki/is-loki-running';
import type {PlayFunction} from '@storybook/csf';
// @ts-expect-error
import type {ReactFramework} from '@storybook/react';

async function waitForDocumentLoaded(): Promise<void> {
  if (document.readyState === 'loading') {
    return new Promise(resolve => {
      document.addEventListener('DOMContentLoaded', () => resolve());
    });
  }

  return Promise.resolve();
}

function enablePointerEvents(): void {
  const styleElement = document.createElement('style');
  document.documentElement.appendChild(styleElement);
  styleElement.sheet?.insertRule('* {pointer-events: auto !important}');
}

export function lokiPlayDecorator(
  target: PlayFunction<ReactFramework>
): PlayFunction<ReactFramework> {
  return async context => {
    const lokiRunning = isLokiRunning();
    const asyncCallback = createAsyncCallback();

    if (lokiRunning) {
      await waitForDocumentLoaded();
      enablePointerEvents();
    }

    try {
      await target(context);
    } finally {
      asyncCallback();
    }
  };
}
