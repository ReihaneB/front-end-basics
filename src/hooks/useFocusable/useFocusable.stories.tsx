import { useRef } from 'react';
import { userEvent, within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

import { lokiPlayDecorator } from '@/testUtils/loki/play-decorator';

import { useFocusable } from './useFocusable';

import styles from './useFocusable.stories.module.css';

export default {
  title: 'Hooks/UseFocusable',
  component: useFocusable,
  argTypes: {
    autoFocus: {
      control: {
        type: 'boolean',
      },
    },
  },
};

function FocusableComponent({
  autoFocus = false,
  buttonTestId = 'focus-button',
  focusableTestId = 'focusable',
}) {
  const ref = useRef<HTMLElement | null>(null);

  const { ref: elementRef, setFocus } = useFocusable(ref, { autoFocus });

  return (
    <div>
      <div
        className={styles.focusable}
        role="button"
        tabIndex={0}
        ref={elementRef}
        data-testid={focusableTestId}
      >
        Click the button to focus it. This element is
      </div>
      <button
        type="button"
        onClick={setFocus}
        data-testid={buttonTestId}
      >
        Click me to focus
      </button>
    </div>
  );
}

export function Playground({...args}) {
  return <FocusableComponent {...args} />;
}

Playground.args = {
  autoFocus: false,
};

export function AutoFocus() {
  return <FocusableComponent autoFocus />;
}

export function ClickToFocus() {
  return <FocusableComponent />;
}

ClickToFocus.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('focus-button'));

    expect(canvas.getByTestId('focusable')).toHaveFocus();
  }
);

export function FocusOnTab() {
  return (
    <FocusableComponent />
  );
}

FocusOnTab.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.tab();

    expect(canvas.getByTestId('focusable')).toHaveFocus();
  }
);

export function FocusAssociated() {
  return (
    <div>
      <FocusableComponent buttonTestId="focus-button-1" focusableTestId="focusable-1" />
      <FocusableComponent buttonTestId="focus-button-2" focusableTestId="focusable-2" />
    </div>
  );
}

FocusAssociated.storyName = 'Focus the correct element associated with its ref on click';

FocusAssociated.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('focus-button-2'));

    expect(canvas.getByTestId('focusable-2')).toHaveFocus();
  }
);
