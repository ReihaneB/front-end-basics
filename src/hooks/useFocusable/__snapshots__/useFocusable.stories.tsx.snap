// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks/UseFocusable AutoFocus smoke-test 1`] = `
<div>
  <div class="src-hooks-useFocusable-useFocusable-stories-module__focusable--CKkxg src-hooks-useFocusable-useFocusable-stories-module__focus--l8_mi"
       role="button"
       tabindex="0"
       data-testid="focusable"
  >
    Click the button to focus it. This element is
  </div>
  <button type="button"
          data-testid="focus-button"
  >
    Click me to focus
  </button>
</div>
`;

exports[`Hooks/UseFocusable ClickToFocus play-test 1`] = `
<div>
  <div class="src-hooks-useFocusable-useFocusable-stories-module__focusable--CKkxg src-hooks-useFocusable-useFocusable-stories-module__focus--l8_mi"
       role="button"
       tabindex="0"
       data-testid="focusable"
  >
    Click the button to focus it. This element is
  </div>
  <button type="button"
          data-testid="focus-button"
  >
    Click me to focus
  </button>
</div>
`;

exports[`Hooks/UseFocusable FocusAssociated play-test 1`] = `
<div>
  <div>
    <div class="src-hooks-useFocusable-useFocusable-stories-module__focusable--CKkxg src-hooks-useFocusable-useFocusable-stories-module__focus--l8_mi"
         role="button"
         tabindex="0"
         data-testid="focusable-1"
    >
      Click the button to focus it. This element is
    </div>
    <button type="button"
            data-testid="focus-button-1"
    >
      Click me to focus
    </button>
  </div>
  <div>
    <div class="src-hooks-useFocusable-useFocusable-stories-module__focusable--CKkxg src-hooks-useFocusable-useFocusable-stories-module__focus--l8_mi"
         role="button"
         tabindex="0"
         data-testid="focusable-2"
    >
      Click the button to focus it. This element is
    </div>
    <button type="button"
            data-testid="focus-button-2"
    >
      Click me to focus
    </button>
  </div>
</div>
`;

exports[`Hooks/UseFocusable FocusOnTab play-test 1`] = `
<div>
  <div class="src-hooks-useFocusable-useFocusable-stories-module__focusable--CKkxg src-hooks-useFocusable-useFocusable-stories-module__focus--l8_mi"
       role="button"
       tabindex="0"
       data-testid="focusable"
  >
    Click the button to focus it. This element is
  </div>
  <button type="button"
          data-testid="focus-button"
  >
    Click me to focus
  </button>
</div>
`;

exports[`Hooks/UseFocusable Playground smoke-test 1`] = `
<div>
  <div class="src-hooks-useFocusable-useFocusable-stories-module__focusable--CKkxg src-hooks-useFocusable-useFocusable-stories-module__focus--l8_mi"
       role="button"
       tabindex="0"
       data-testid="focusable"
  >
    Click the button to focus it. This element is
  </div>
  <button type="button"
          data-testid="focus-button"
  >
    Click me to focus
  </button>
</div>
`;
